### PR TITLE
Add compile error on const fn instrumentation

### PR DESCRIPTION
**Pull request by andrewpollack**
Fixes: #2414

I'm newer to using macro expansions to enforce `compile_error!` calls, and this is my interpretation on how to approach it.  Unfortunately, the `compile_error!` is triggering regardless of whether I'm instrumenting a function or not, which makes me think I'm not approaching the code generation portion correctly.

Should I be forming this as a `proc_macro` instead of a `macro_rules`?

### CODE REVIEW

**Review by hawkw — COMMENTED** [review]
The code you've written expands the `compile_error!` macro inside the implementation of the `instrument` proc macro, rather than including the `compile_error!` macro inside the _output_ of the `instrument` proc macro. This means that the error occurs when _compiling_ the `tracing-attributes` crate, because the `compile_error!` macro is present in its source.

In order for this to work correctly, the `compile_error!` macro needs to be generated in the `instrument` proc macro's _output_. Proc macros like `instrument` typically use the [`quote!` macro](https://docs.rs/quote/latest/quote/) to emit generated code.  To illustrate this, note that whenever we use `compile_error!` elsewhere in `tracing-attributes`, it is always inside of `quote!`d code that will be output by the `instrument` macro: https://github.com/tokio-rs/tracing/blob/02903cbcdcbec946fd4086395409b1cacd9807de/tracing-attributes/src/expand.rs#L168-L170

You can fix this by changing your branch to return a `quote!`d `compile_error!` when encountering a `const fn`.

**Comment by andrewpollack** [comment]
@hawkw your explanation was fantastic and makes tons of sense, thank you so much! 😸

I've pushed the `compile_error` update, along with an additional doc test and a couple tests to `tracing-attributes/tests/ui` (I wasn't sure if these were the best places to put those updates, so I'm happy to update if there are better locations!)

File: tracing-attributes/src/lib.rs [hunk 1]
```diff
@@ -385,6 +393,13 @@ fn instrument_precise(
     let input = syn::parse::<ItemFn>(item)?;
     let instrumented_function_name = input.sig.ident.to_string();
 
+    if input.sig.constness.is_some() {
+        return Ok(quote! {
+            compile_error!("instrumenting const fn not supported")
```
Comment (hawkw) [inline tracing-attributes/src/lib.rs:398]: this feels a little bit terse; what do you think about rephrasing this as 
```suggestion
            compile_error!("the `#[instrument]` attribute may not be used with `const fn`s")
```

**Review by hawkw — APPROVED** [review]
overall, this looks good to me! i had a small suggestion regarding the wording of the error message.

File: tracing-attributes/src/lib.rs (hunk 1 above)
Comment (andrewpollack) [reply tracing-attributes/src/lib.rs]: Great clarity -- thank you!

**Review by hawkw — APPROVED** [review]
this looks great to me, and i'll merge it as soon as CI passes! thanks for working on this!

**Pull request merged**